client/history/peer_resolver.go

```diff
@@ -110,3 +115,54 @@ func (pr peerResolver) GetAllPeers() ([]string, error) {
 	}
 	return peers, nil
 }
+
+func (pr peerResolver) GlobalRatelimitPeers(ratelimits []string) (map[string][]string, error) {
+	// History was selected simply because it already has a ring and an internal-only API.
+	// Any service should be fine, it just needs to be shared by both ends of the system.
+	hosts, err := pr.resolver.Members(service.History)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get history peers: %w", err)
+	}
+	if len(hosts) == 0 {
+		// can occur when shutting down the only instance because it calls EvictSelf ASAP.
+		// this is common in dev, but otherwise *probably* should not be possible in a healthy system.
+		return nil, fmt.Errorf("unable to get history peers: no peers available")
+	}
+
+	results := make(map[string][]string, len(hosts))
+	initialCapacity := len(ratelimits) / len(hosts)
+	// add a small buffer to reduce copies, as this is only an estimate
+	initialCapacity += initialCapacity / 10
+	// but don't use zero, that'd be pointless
+	if initialCapacity < 1 {
+		initialCapacity = 1
+	}
+
+	for _, r := range ratelimits {
+		// figure out the destination for this ratelimit
+		host, err := pr.resolver.Lookup(service.History, r)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"unable to find host for ratelimit key %q: %w", r, err)
+		}
+		peer, err := host.GetNamedAddress(pr.namedPort)
+		if err != nil {
+			// AFAICT should not happen unless the peer data is incomplete / the ring has no port info.
+			// retry may work?  eventually?
+			return nil, fmt.Errorf(
+				"unable to get address from host: %s, %w",
+				host.String(), // HostInfo has a readable String(), it's good enough
+				err,
+			)
+		}
+
+		// add to the peer's partition
+		current := results[peer]
+		if len(current) == 0 {
```

Comment:
tbh I don't think I'd label that "more readable" - it allows a third category of behavior, where "current" exists but is empty.  `len(current)==0` has only two cases to consider: empty or not.

Reply:
I added as a nit comment but since you are also interested in this topic let me share more on my reasoning. (not a blocker for this PR by the way)

Both options achieve the same thing for this case. The only advantage of using `if !ok` approach is the consistency across various map existence check scenarios in the codebase. I guess we'd agree that when a convention is consistently applied in the codebase it helps with reading experience. If each developer chooses a viable option for their own taste then you lose some readability score.
Bool ok is part of the map interface and handles all cases. However using the values to determine existence fall short if the map can contain values that are the default values such as integer 0 or a zero struct value. In that case checking `if myMap[key] != 0` doesn't work. This means you would have to use `if !ok` for some maps and zero value check for others. Obviously inconsistent => less readable.

We can argue how many milliseconds does this add to reading experience :) From principles perspective, I see the obvious choice of sticking with the approach that works for all cases to stay consistent.

Reply:
from a filtered search of the codebase (excluding tests and mocks and .gen), we have roughly:
- 336 uses of `val, ok := something[key]`
- 1,886 uses of `val := something[key]`

and yeah, zero values are valid and safe in this case, hence why I don't differentiate between "exists" and "not exists".  if they were unsafe, I'd absolutely agree, but that's not the case here.